src/context/TodoContext.tsx

```diff
@@ -0,0 +1,149 @@
+import React, {
+  MutableRefObject,
+  ReactNode,
+  SetStateAction,
+  createContext, useContext, useEffect, useRef, useState,
+} from 'react';
+import { editTodo, getTodos } from '../api/todos';
+import { SortTypes, Todo } from '../types/Todo';
+
+export interface TContext {
+  todos: Todo[];
+  setTodos: React.Dispatch<SetStateAction<Todo[]>>;
+  hasError: null | string;
+  handleError: (error: string) => void;
+  sortType: SortTypes;
+  setSortType: React.Dispatch<SetStateAction<SortTypes>>;
+  setHasError: React.Dispatch<SetStateAction<string | null>>;
+  tempTodos: Todo | null;
+  setTempTodos: React.Dispatch<SetStateAction<Todo | null>>;
+  handleToggleStatus: (todoId: number) => void ;
+  isToggled: boolean,
+  setIsToggled: React.Dispatch<SetStateAction<boolean>>,
+  handleToggleAllStatus:() => void,
+  isToggledAll: boolean,
+  setIsToggledAll: React.Dispatch<SetStateAction<boolean>>,
+  titleInputRef: MutableRefObject<HTMLInputElement | null>,
+  editedRef: MutableRefObject<HTMLInputElement | null>,
+  isGroupDeleting: boolean,
+  setIsGroupDeleting: React.Dispatch<SetStateAction<boolean>>,
+}
+
+const TodoContext = createContext<TContext | null>(null);
+
+export function useTodoContext() {
+  return useContext(TodoContext);
+}
+
+const USER_ID = 11550;
+
+export function TodoProvider({ children }: { children: ReactNode }) {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [hasError, setHasError] = useState<string | null>(null);
```

Comment:
Let's narrow down the type of this state to union of strings which we are actually using there + null